### ISSUE TITLE
combinators and typeclass instances on Argument

### ISSUE THREAD

**Issue by johnynek** [bkirwi/decline#182]
We could definitely have `Apply[Argument]` it seems (we don't have pure because can't supply a defaultMetavar) if we take the "left most" or "right most" defaultMetavar.

We can certainly have `Functor[Argument]` even without the above constraint.

But if we would allow the above (say take the left side). I think we can also have `Alternative[Argument]` which can be useful to compose different Arguments.

One thing I like is an
```scala
def either[A, B](a: Argument[A], b: Argument[B]): Argument[Either[A, B]] =
  Alternative[Argument].combineK(b.map(Right(_)), a.map(Left(_))) 
```

This allows a single flag like `--input` to parse multi values (say uris but also local file paths)

**Comment by bkirwi** [comment]
> One thing I like is an `def either[A, B](a: Argument[A], b: Argument[B]): Argument[Either[A, B]] = ...`

I also like this! Could even make sense to have it in implicit scope.

> I think we can also have Alternative[Argument] which can be useful to compose different Arguments.

The zero value gets a bit annoying here two -- what's `Alternative.empty` return? I also worry a bit about what `ap` returns... the obvious thing to do is to parse the same string with both parsers, but that doesn't strike me as super useful in this context, and we don't have any standard delimiter we can use to split...

Basically: enthusiastic about the `either` combinator but a bit more hesitant on the general stuff. Happy to reconsider if you have more cases in mind where it would be useful though!

**Comment by johnynek** [comment]
I'm not sure Argument.nothing or something is that weird. Note you do have a similar Opts.

Just have read return `Validated.invalidNel("Argument.nothing matches no input")` and have some sensible defaultMetavar.

If we did that we could have a MonoidK rather than a SemigroupK.

Note, users are free to define this on their own as well.

**Comment by johnynek** [comment]
But I agree Applicative can be defined but it's a bit unintuitive to parse the same input two different ways.

I can't see a lot of value in that. combineK seems valuable as does map.
